File: src/python/demucs_wrapper.py
#!/usr/bin/env python3
"""
Demucs Wrapper - IPC interface for music source separation

Reads commands from stdin and outputs JSON progress messages to stdout.
"""

import json
import sys
import os
from pathlib import Path

try:
    import torch
    from demucs.separate import Separator
    from demucs.pretrained import model_name_to_metadata
except ImportError as e:
    print(json.dumps({
        "type": "error",
        "message": f"Failed to import required modules: {str(e)}"
    }))
    sys.exit(1)


class DemucsIPCInterface:
    def __init__(self):
        self.separator = None
        self.model_name = "htdemucs"
        
    def log_progress(self, percent, stage, model, current_chunk, total_chunks):
        """Send progress update to parent process via JSON"""
        msg = {
            "type": "progress",
            "percent": percent,
            "stage": stage,
            "model": model,
            "current_chunk": current_chunk,
            "total_chunks": total_chunks
        }
        print(json.dumps(msg), flush=True)
    
    def log_error(self, message):
        """Send error message to parent process via JSON"""
        msg = {
            "type": "error",
            "message": message
        }
        print(json.dumps(msg), flush=True)
    
    def initialize(self):
        """Initialize Demucs model"""
        try:
            self.log_progress(10, "Initializing model", self.model_name, 0, 1)
            
            # Create separator with appropriate device
            device = "cuda" if torch.cuda.is_available() else "cpu"
            self.separator = Separator(
                model=self.model_name,
                device=device,
                segment=30,  # Process in 30-second chunks
                overlap=0.1,
                shifts=1
            )
            
            self.log_progress(20, "Model loaded", self.model_name, 0, 1)
            return True
        except Exception as e:
            self.log_error(f"Failed to initialize model: {str(e)}")
            return False
    
    def separate(self, input_file, output_folder, stems):
        """Separate audio file into stems"""
        try:
            if not self.separator:
                if not self.initialize():
                    return False
            
            input_path = Path(input_file)
            if not input_path.exists():
                self.log_error(f"Input file not found: {input_file}")
                return False
            
            output_path = Path(output_folder)
            output_path.mkdir(parents=True, exist_ok=True)
            
            self.log_progress(30, "Loading audio file", self.model_name, 0, 1)
            
            # Load and separate
            origin, separated = self.separator.separate_audio_file(input_file)
            
            self.log_progress(80, "Saving stems", self.model_name, 0, 1)
            
            # Save separated stems
            saved_count = 0
            for stem, source in separated.items():
                output_file = output_path / f"{stem}.wav"
                from demucs.api import save_audio
                save_audio(source, str(output_file), 
                          samplerate=self.separator.samplerate)
                saved_count += 1
                self.log_progress(80 + (saved_count * 2), "Saving stems", self.model_name, saved_count, len(separated))
            
            self.log_progress(100, "Separation complete", self.model_name, len(separated), len(separated))
            return True
        
        except Exception as e:
            self.log_error(f"Separation failed: {str(e)}")
            return False


def main():
    """Main entry point - reads arguments from command line"""
    if len(sys.argv) < 3:
        print(json.dumps({
            "type": "error",
            "message": "Usage: demucs_wrapper.py <input_file> <output_folder> [stems...]"
        }))
        sys.exit(1)
    
    input_file = sys.argv[1]
    output_folder = sys.argv[2]
    stems = sys.argv[3:] if len(sys.argv) > 3 else ["vocals", "drums", "bass", "other"]
    
    interface = DemucsIPCInterface()
    
    # Set up signal handlers for graceful shutdown
    import signal
    def signal_handler(sig, frame):
        print(json.dumps({
            "type": "cancelled",
            "message": "Separation cancelled by user"
        }))
        sys.exit(0)
    
    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)
    
    # Perform separation
    if interface.separate(input_file, output_folder, stems):
        sys.exit(0)
    else:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/python/demucswrapper.cpp
#include "demucswrapper.h"
#include "pythonprocess.h"
#include "../utils/logger.h"
#include <QJsonObject>
#include <QJsonArray>

DemucsWrapper::DemucsWrapper(const QString &pythonPath, const QString &wrapperScriptPath)
    : pythonPath_(pythonPath)
    , wrapperScriptPath_(wrapperScriptPath)
{
    process_ = std::make_unique<PythonProcess>(pythonPath, wrapperScriptPath);

    process_->setProgressCallback([this](const QJsonObject &data) {
        onProgress(data);
    });

    process_->setErrorCallback([this](const QString &message) {
        onError(message);
    });

    process_->setCompletedCallback([this]() {
        onCompleted();
    });
}

DemucsWrapper::~DemucsWrapper()
{
    if (isRunning()) {
        cancel();
    }
}

bool DemucsWrapper::separate(const QString &inputFile,
                             const QStringList &stems,
                             const QString &outputFolder)
{
    if (isRunning()) {
        Logger::log(Logger::Warning, "Demucs separation already running");
        return false;
    }

    Logger::log(Logger::Info, QString("Starting Demucs separation: %1").arg(inputFile));
    Logger::log(Logger::Info, QString("Stems: %1").arg(stems.join(", ")));
    Logger::log(Logger::Info, QString("Output: %1").arg(outputFolder));

    // Build command line arguments
    QStringList args;
    args << inputFile;
    args << outputFolder;

    // Add stems as arguments
    for (const auto &stem : stems) {
        args << stem.toLower();
    }

    return process_->start(args);
}

void DemucsWrapper::cancel()
{
    if (isRunning()) {
        Logger::log(Logger::Info, "Canceling Demucs separation...");
        process_->stop();
    }
}

bool DemucsWrapper::isRunning() const
{
    return process_ && process_->isRunning();
}

void DemucsWrapper::onProgress(const QJsonObject &data)
{
    QString type = data["type"].toString();

    if (type == "progress") {
        SeparationProgress progress;
        progress.percent = data["percent"].toInt();
        progress.stage = data["stage"].toString();
        progress.model = data["model"].toString();
        progress.currentChunk = data["current_chunk"].toInt();
        progress.totalChunks = data["total_chunks"].toInt();

        Logger::log(Logger::Debug, QString("Progress: %1% - %2 (%3/%4)").arg(progress.percent).arg(progress.stage).arg(progress.currentChunk).arg(progress.totalChunks));

        if (progressCallback_) {
            progressCallback_(progress);
        }
    } else if (type == "error") {
        QString errorMsg = data["message"].toString();
        Logger::log(Logger::Error, QString("Demucs error: %1").arg(errorMsg));
        onError(errorMsg);
    } else {
        Logger::log(Logger::Debug, QString("Demucs message: %1").arg(type));
    }
}

void DemucsWrapper::onError(const QString &message)
{
    if (errorCallback_) {
        errorCallback_(message);
    }
}

void DemucsWrapper::onCompleted()
{
    Logger::log(Logger::Info, "Demucs separation completed");
    // TODO: Get output folder from Python process result
    if (completedCallback_) {
        completedCallback_(QString());
    }
}
